src/operator/mshadow_op.h

```diff
@@ -1532,6 +1551,71 @@ struct sum {
   }
 };
 
+/*! \brief arg max reducer */
+struct argmax {
+  /*! \brief do reduction into dst */
+  template<typename AType, typename DType>
+  MSHADOW_XINLINE static void Reduce(volatile AType& dst,  volatile DType src) { // NOLINT(*)
+    if (dst.num < src.num) {
+      dst.num = src.num;
+      dst.idx = src.idx;
+    }
+  }
+  /*! \brief do stable reduction into dst */
+  template<typename AType, typename DType>
+  MSHADOW_XINLINE static void Reduce(volatile AType& dst,  volatile DType src, volatile DType& residual) { // NOLINT(*)
```

Comment:
purpose of residual ?

Reply:
residual is not used in argmin/max. When doing sum it's useful for safe accumulation. For compatibility with the reduce kernel we need to define this function with residue even if it's not used